### PR TITLE
Shehadak/earley parser

### PR DESCRIPTION
**Pull request by shehadak**
# Model Addition: Probabilistic Earley Parser

Added a probabilistic Earley parser implementation (Stolcke, 1995), building on top NLTK's built-in incremental chart parsing implementation. Configured the implementation to be compatible with the Artificial Subject interface and flexible with different forms of grammar.

## This PR
This PR contains:
* `utils.py`: contains definitions for chart rules, adapted from NLTK's chart implementation but modified to support incremental chart parsing with a probabilistic grammar.
* `parser.py`: contains the bulk of the model implementation, including the definition of `EarleyParserSubject`.
* `test.py`: two sets of unit tests, each testing the behavioral tasks on different grammars, and one integration test targeting the end-to-end model loading and text processing.
* `__init__.py`: registers the model.
* `treebank/sample_treebank`: a toy treebank of manually-annotated sentences. Currently, this treebank is being used to generate the default grammar. However, depending on the use case, a user might elect to add their own grammar file and load it instead (e.g. PTB).
* `requirements.txt`

## Current Features
* Grammar flexibility: the current implementation allows users to either pass a grammar string, use one of NLTK's corpora to construct a grammar, or place their own treebank corpus in the `treebank` directory and construct a grammar using that corpora instead. This was particularly done to support loading PTB and other preprocessed treebanks locally without having to publish the models publicly. 
* Unkification: when creating a grammar from a treebank, the user may elect to unkify (i.e. replace with `<unk>` token) low-frequency words in the corpus before constructing the grammar. This is particularly helpful as it is difficult to attain a universal grammar that works for all benchmark datasets, and a grammar that learns the `<unk>` token eases that restriction.

### Reading Times
* The model computes the probability of a partial parse by calculating a prefix probability $\alpha_n$ for word $n$. For the first word, $\alpha_0$ is the marginal probability over all potential parses. For additional words, the prefix probability is determined as the sum of inside probabilities of each possible parse weighted by the prefix probability of that parse. The probability of a partial parse, therefore, is the front probability ending at the last observed terminal in that parse.
* The model uses surprisal (in bits) as a proxy for reading times, following (Smith & Levy 2013). Surprisal is calculated cumulatively as $-log_2 P(\alpha_i)$. This is consistent with (Hale, 2001), but the difference here is that I compute surprisal cumulatively instead of per word.

### Next Word Prediction
* The model predicts the next word given a partial parse by identifying the most likely grammar production in the current chart step that is 1) incomplete (i.e. hasn't finished parsing), and 2) just completed the last observed terminal. Then, it identifies the following node in that production and recursively chooses its most likely production until it reaches a terminal node. The model therefore deterministically returns the word that maximizes the probability $P(x_{t+1} | x_{\leq t})$.
* The implementation therefore always assumes that the parse is partial and incomplete. This assumption is valid as users can choose to have explicit markers for completion (e.g., a full stop production `. -> '.'`).

## Limitations
* Neural tasks are not supported as this is a behavioral model.
* Runtime: the current implementation is extremely slow for large grammars. For context, a parser 
trained on %10 of PTB takes longer than 16 hours to parse `the quick brown fox jumps over the lazy dog`. The StanfordNLP parser speeds up parsing by using an unlexicalized parser (Klein & Manning, 2003) and using FSA minimization.
* When calling `digest_text` with a list of strings, the current implementation requires that the terminals contained in each part be a single syntactic unit. This surely works when the entire string is passed `"The quick brown fox jumps over the lazy dog` and when each word is passed separately `["The", "quick", "brown", "fox",  "jumps", "over", "the", "lazy", "dog]"`. However, if an input such as `["The quick brown", "fox jumps over", "the lazy dog"]`, the parser fails as `"fox jumps over"` combines non-terminals from an NP and a VP (assuming a grammar that uses `S -> NP VP`). This is a limitation of using chart parsing, and additional work is needed to come up with a working solution.

## References
[The effect of word predictability on reading time is logarithmic
](https://www.sciencedirect.com/science/article/pii/S0010027713000413) (Smith & Levy, _Cognition_ 2013)
[An efficient probabilistic context-free parsing algorithm that computes prefix probabilities](https://arxiv.org/pdf/cmp-lg/9411029.pdf) (Stolcke, _arXiv_ 1994)
[A Probabilistic Earley Parser as a Psycholinguistic Model](https://aclanthology.org/N01-1021.pdf) (Hale, _ACL_ 2001)
[Accurate Unlexicalized Parsing](https://aclanthology.org/P03-1054) (Klein & Manning, _ACL_ 2003)

### CODE REVIEW

File: brainscore_language/models/earley_parser/parser.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+class EarleyParserSubject(ArtificialSubject):
+    def __init__(
+        self,
+        model_id: str,
+    ):
+        """
+        :param model_id: the model id i.e. name
+        """
+        self._logger = logging.getLogger(fullname(self))
+        self.model_id = model_id
```
Comment (mschrimpf) [inline brainscore_language/models/earley_parser/parser.py:32]: since the `model_id` doesn't affect anything else in constructing the object, I would hardcode:
```suggestion
    ):
        self._logger = logging.getLogger(fullname(self))
        self.model_id = 'earley-parser'
```

Unless you want the grammar to be dynamic too, then I suppose the `treebank_path` should be an argument along with the `model_id`

File: brainscore_language/models/earley_parser/parser.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+            xr.concat(output["behavior"], dim="presentation").sortby("part_number")
+            if output["behavior"]
+            else None
+        )
+
+        return output
+
+    def estimate_reading_times(self, chart: Chart, start: int, end: int):
+        """
+        :param parses: a list of probabilistic edges from the parsing chart
```
Comment (mschrimpf) [inline brainscore_language/models/earley_parser/parser.py:134]: doc doesn't match function signature

File: brainscore_language/models/earley_parser/parser.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+        self._logger.debug("Merging outputs")
+        output["behavior"] = (
+            xr.concat(output["behavior"], dim="presentation").sortby("part_number")
+            if output["behavior"]
+            else None
+        )
+
+        return output
+
+    def estimate_reading_times(self, chart: Chart, start: int, end: int):
```
Comment (mschrimpf) [inline brainscore_language/models/earley_parser/parser.py:132]:
```suggestion
    def estimate_reading_times(self, chart: Chart, start: int, end: int) -> float:
```
is this correct?

File: brainscore_language/models/earley_parser/parser.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+            (https://www.sciencedirect.com/science/article/pii/S0010027713000413)
+        """
+        # Parse context and get parses
+        edges = chart.select(end=end)
+
+        # Exclude edges from predict rules
+        edges = [p for p in edges if not p.start() == p.end()][1:]
+
+        if not edges:
+            # Could not parse the prefix, edge case --> infinte surprisal
```
Comment (mschrimpf) [inline brainscore_language/models/earley_parser/parser.py:145]:
```suggestion
            # Could not parse the prefix, edge case --> infinite surprisal
```

File: brainscore_language/models/earley_parser/test.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+
+
+@pytest.mark.parametrize(
+    "model_identifier, treebank_path, expected_reading_times, expected_next_words",
+    [
+        (
+            "earley-parser",
+            str(Path(__file__).parent / "treebank"),
+            [1.15200, 3.4739, 6.2109, 8.08537, 9.4073, 10.5593, 12.5593, 13.8812],
+            ["lazy", "man", "saw", "with", "a", "lazy", "man", "<unk>"],
```
Comment (mschrimpf) [inline brainscore_language/models/earley_parser/test.py:143]: I would prefer to separate this into two tests, one for reading times and one for next words. This might be personal preference (I like detailed test reports), so if you think this is the way to go, feel free to ignore.

**Review by mschrimpf — CHANGES_REQUESTED** [review]
can't speak to the correctness of the parser/grammar, but overall model implementation looks good to me, subject to small tweaks.

**Comment by mschrimpf** [comment]
@hans This model implementation with the `sample_treebank` is very simple and will likely lead to a poor score for `earley-parser`. I'm okay with this, but if there is a more expressive runnable grammar that you would like to see (or that the community expects / standardly evaluates), please send in a model with that specification (or ping someone who can help).

**Comment by hans** [comment]
> @hans This model implementation with the `sample_treebank` is very simple and will likely lead to a poor score for `earley-parser`. I'm okay with this, but if there is a more expressive runnable grammar that you would like to see (or that the community expects / standardly evaluates), please send in a model with that specification (or ping someone who can help).

My impression is that anything more expressive (e.g. the grammar learned on PTB) is going to be extremely compute-limited with this implementation. So I think the sample treebank may have to do for now, unless/until we have the dev cycles to import something more scalable (e.g. [`earleyx`](https://github.com/lmthang/earleyx)).

**Review by mschrimpf — APPROVED** [review]
